### DISCUSSION TITLE
YoloV5x-high prediction time

### DISCUSSION THREAD

**Discussion by revathib01** [ultralytics/yolov5#5703] [Q&A]
Hi,

I have trained my custom dataset of 300 images using all 4  versions of yolov5 with image size 640 in Google Colab (selected YoloV5x based on the initial prediction accuracy) . We tested the Yolov5x model accuracy in both Google Colab and a local Pc with Ubuntu operating system for Image sizes  640 and 320. The time taken for prediction in the  local PC is high. Below is the comparison between the Google Colab and local PC,

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/DELL/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/DELL/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

Platform | Runtime | Image size | Time Taken per   image
-- | -- | -- | --
Google   Colab | Gpu | 640 | 0.1-0.2 secs
Google   Colab | Cpu | 640 | 2.7-3.6 secs
Local Linux   Pc | Cpu | 640 | 7.5-12.5secs
Google   Colab | Gpu | 320 | 0.07-0.11secs
Google   Colab | Cpu | 320 | 0.8-1secs
Local Linux   Pc | Cpu | 320 | 2.0-3.6 secs

</body>

</html>

I want to deploy my final model on a single board computer but, as you can see from above table prediction time is high. Based on the accuracy of the  model we would like to go with Yolov5x model. From the above table we found we may need specific hardware requirement to run the model where the prediction can take time between 100-200ms.

Can you please let us know the Hardware requirements that meet my requirements with yolov5x model (Image Size 640).

**Comment by glenn-jocher** [comment]
@revathib01 👋 Hello! Thanks for asking about **inference speed issues**. YOLOv5 🚀 can be run on CPU (i.e. `--device cpu`, slow) or GPU if available (i.e. `--device 0`, faster). You can determine your inference device by viewing the YOLOv5 console output:

### detect.py inference
```bash
python detect.py --weights yolov5s.pt --img 640 --conf 0.25 --source data/images/
```
<img width="944" alt="detect.py" src="https://user-images.githubusercontent.com/26833433/140100842-e130b5f7-64cb-46c9-9386-72abadd19828.png">

### YOLOv5 PyTorch [Hub](https://github.com/ultralytics/yolov5/issues/36) inference
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')

# Images
dir = 'https://ultralytics.com/images/'
imgs = [dir + f for f in ('zidane.jpg', 'bus.jpg')]  # batch of images

# Inference
results = model(imgs)
results.print()  # or .show(), .save()
# Speed: 631.5ms pre-process, 19.2ms inference, 1.6ms NMS per image at shape (2, 3, 640, 640)
```

### Increase Speeds

If you would like to increase your inference speed some options are:

- Use batched inference with [YOLOv5 PyTorch Hub](https://github.com/ultralytics/yolov5/issues/36)
- Reduce `--img-size`, i.e. 1280 -> 640 -> 320
- Reduce model size, i.e. YOLOv5x -> YOLOv5l -> YOLOv5m -> YOLOv5s -> YOLOv5n
- Use half precision FP16 inference with `python detect.py --half` and `python val.py --half`
- Use a faster GPUs, i.e.: P100 -> V100 -> A100
- [Export](https://github.com/ultralytics/yolov5/issues/251) to ONNX for 2-3x CPU speedup with detect.py
- Use a free GPU backends with up to 16GB of CUDA memory: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
